### PR TITLE
Data type "color"

### PR DESCRIPTION
**Pull request by nhaagen**
Introduces the data-type "color". 
Color can be constructed using a regular or short hex-value as well as RGB-values.
Same for value retrieval.

### CODE REVIEW

File: src/Data/Color/Color.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	public function rgbstring() {
+		return 'rgb('
+			.implode($this->rgb(), ', ')
+			.')';
+	}
+
+	/**
+	 * Return a list with rgb-values of color
+	 *
+	 * @return <int>
```
Comment (klees) [inline src/Data/Color/Color.php]: `int[]` or `array<int>` for an array of integer.

File: src/Data/Color/Color.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ * in the RGB color space.
+ *
+ * @author Nils Haagen <nils.haagen@concepts-and-training.de>
+ */
+class Color {
+
+	/**
+	 * @var string
+	 */
+	protected $value;
```
Comment (klees) [inline src/Data/Color/Color.php]: Store as `$r`, `$g` and `$b` and `__construct` accordingly.

File: src/Data/Color/Color.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	public function __construct($value) {
+		$this->value = $this->normalize($value);
+	}
+
+	/**
+	 * Return the hex-value of color
+	 *
+	 * @return string
+	 */
+	public function value() {
```
Comment (klees) [inline src/Data/Color/Color.php]: This will e.g. return `#03a417`? Maybe this should be named `asHex`?

File: src/Data/Color/Color.php [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	/**
+	 * Return string with rgb-notation
+	 *
+	 * @return string
+	 */
+	public function rgbstring() {
+		return 'rgb('
+			.implode($this->rgb(), ', ')
+			.')';
+	}
```
Comment (klees) [inline src/Data/Color/Color.php]: Maybe add individual `r`, `g` and `b` getter? Maybe use `as*`for all other formats (rgb-string, hex, array)?

File: src/Data/Color/Color.php [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	 *
+	 * @return <int>
+	 */
+	public function rgb() {
+		$col = $this->trimhash($this->value);
+		$chunks = str_split($col, 2);
+		return array_map('hexdec', $chunks);
+	}
+
+	/**
```
Comment (klees) [inline src/Data/Color/Color.php]: Short description missing. `@return` missing.

File: src/Data/Color/Color.php [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+				if(count($value) !== 3) {
+					throw new \UnexpectedValueException("RGB must contain three values", 1);
+				}
+				// build constraints
+				$f = new VAL\Factory(new D\Factory());
+				$constraints = $f->parallel([
+						$f->greaterThan(-1),
+						$f->lessThan(256)
+					],
+					new D\Factory()
```
Comment (klees) [inline src/Data/Color/Color.php]: This is is wrong here.

File: src/Data/Color/Color.php [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	 * -is a list of exactly three values
+	 * -each value is between 0 and 255
+	 *
+	 * The datatype stores the value in hex longhand notation.
+	 * Normalization also includes validation.
+	 *
+	 * @throws Exception
+	 * @return string
+	 */
+	private function normalize($value) {
```
Comment (klees) [inline src/Data/Color/Color.php]: Move all this specific cases of how a value could be inputted to a new `ColorFactory`. Give them names (like "fromHex", "fromRGB", ...) and split them apart. Make central `build`-method, that decides which `from`-method to use. I.e.: refactor this =)

**Review by klees — CHANGES_REQUESTED** [review]
Change storage and creation please. Documentation is thorough. Thx.

**Comment by Amstutz** [comment]
@klees is this ready to be merged? Do we need a JF discussion for this?

**Comment by klees** [comment]
@Amstutz I have to rereview, but it should be mostly fine. I don't see the necessity to JF this, do you see?

File: src/Data/Color/ColorFactory.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	 * a hex-value both short and longhand notation.
+	 *
+	 * @param  string|int[] $value
+	 * @throws \InvalidArgumentException
+	 * @return Color
+	 */
+	public function build($value) {
+
+		if(is_array($value)) {
+			$this->validateRGB($value);
```
Comment (klees) [inline src/Data/Color/ColorFactory.php]: Please change names from "validate*" to "check*". This is a convention we also use in the UI-framework and I try to use in general.

**Review by klees — CHANGES_REQUESTED** [review]
Minor naming, the rest is good. Thx!

**Review by klees — APPROVED** [review]

**Comment by Amstutz** [comment]
Nope and looks good, thanks a lot @nhaagen and @klees

**Review by Amstutz — COMMENTED** [review]
If you find the time, you could enhance the readme with a brand new color example. This is not a requirement however.

**Pull request merged**